File: 916/Fcat_916.go
import "fmt"

func wordSubsets(words1 []string, words2 []string) []string {
    var words2_union = map[rune]int {}
    for _, word2 := range words2 {
        word2_count := map[rune]int {}
        for _, char := range word2 {
            word2_count[char]++
        }
        for key, value := range word2_count{
            words2_union[key] = Max(words2_union[key], value)
        }
    }
    var result [] string
    for _, word1 := range words1 {
        is_subset := true
        word1_count := map[rune]int {}
        for _, char := range word1 {
            word1_count[char]++
        }

        for key, value := range words2_union {
            if value > word1_count[key] {
                is_subset = false
                break
            }
        }
        if is_subset {
            result = append(result, word1)
        }
    }
    return result
}

func Max(x, y int) int {
    if x < y {
        return y
    }
    return x
}

File: 916/Fcat_916.py
from typing import List
from collections import Counter, defaultdict


class Solution:
    def wordSubsets(self, words1: List[str], words2: List[str]) -> List[str]:
        words2_union = defaultdict(int)
        for word2 in words2:
            word2_counter = Counter(word2)
            for char in word2_counter:
                words2_union[char] = max(words2_union[char], word2_counter[char])
        result = []
        for word in words1:
            word_counter = Counter(word)
            for char in words2_union:
                if word_counter.get(char, 0) < words2_union[char]:
                    break
            else:
                result.append(word)
        return result
